Src/Support/GoogleApis.Tests/Apis/Upload/ImprovedResumableUploadTest.cs

```diff
@@ -0,0 +1,966 @@
+        /// <summary>
+        /// Mock string to upload to the media server. It contains 454 bytes, and in most cases we will use a chunk 
+        /// size of 100. There are 3 spaces on the end of each line because the original carriage return line endings
+        /// caused differences between Windows and Linux test results.
+        /// </summary>
+        static readonly string UploadTestData =
+            "Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod   " +
+            "tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris   " +
+            "nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore   " +
+            "eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit   " +
+            "anim id est laborum.";
+        static readonly byte[] uploadTestBytes = Encoding.UTF8.GetBytes(UploadTestData);
+        static readonly int uploadLength = uploadTestBytes.Length;
+
+        /// <summary>
+        /// Stream that doesn't support seeking.
+        /// </summary>
+        private class UnknownSizeMemoryStream : MemoryStream
+        {
+            public UnknownSizeMemoryStream(byte[] buffer) : base(buffer) { }
+            public override bool CanSeek => false;
+            public override long Seek(long offset, SeekOrigin loc)
+            {
+                throw new NotSupportedException();
+            }
+            public override long Position
+            {
+                get { return base.Position; }
+                set { throw new NotSupportedException(); }
+            }
+        }
+
+        /// <summary>
+        /// resumable upload class that allows chunksize to be set for testing.
+        /// </summary>
+        private class MockResumableUpload : ResumableUpload<object>
+        {
+            public MockResumableUpload(IClientService service, string path, string method, Stream stream,
+                string contentType, int chunkSize)
+                : base(service, path, method, stream, contentType)
+            {
+                this.chunkSize = chunkSize;
+            }
+        }
+
+        /// <summary>
+        /// Information about each request made to the server.
+        /// </summary>
+        private class RequestInfo
+        {
+            public RequestInfo(HttpListenerRequest request)
+            {
+                Headers = request.Headers;
+                Url = request.Url;
+            }
+            public NameValueCollection Headers { get; }
+            public Uri Url { get; }
+        }
+
+        /// <summary>
+        /// URL used for the resumable upload.
+        /// </summary>
+        private const string uploadPath = "resume";
+
+        /// <summary>
+        /// HTTP server which listens on localhost:&lt;random port&gt; for testing.
```

Comment:
Possibly add a couple of lines of comment saying that each test registers its own Handler (on a separate URL path) designed to behave in a particular way. Otherwise the server/handler difference takes a little bit of getting used to.